VR/VRInput.cpp

```diff
@@ -0,0 +1,612 @@
+{
+    return (deg * M_PI) / 180.0;
+}
+
+unsigned long sys_timeBase = 0;
+int milliseconds(void)
+{
+	struct timeval tp;
+
+	gettimeofday(&tp, NULL);
+
+	if (!sys_timeBase)
+	{
+		sys_timeBase = tp.tv_sec;
+		return tp.tv_usec/1000;
+	}
+
+	return (tp.tv_sec - sys_timeBase)*1000 + tp.tv_usec/1000;
+}
+
+#ifndef EPSILON
+#define EPSILON 0.001f
+#endif
+
+XrVector3f normalizeVec(XrVector3f vec) {
+    float xxyyzz = vec.x*vec.x + vec.y*vec.y + vec.z*vec.z;
+
+    XrVector3f result;
+    float invLength = 1.0f / sqrtf(xxyyzz);
+    result.x = vec.x * invLength;
+    result.y = vec.y * invLength;
+    result.z = vec.z * invLength;
+    return result;
+}
+
+void GetAnglesFromVectors(const XrVector3f forward, const XrVector3f right, const XrVector3f up, vec3_t angles)
+{
+    float sr, sp, sy, cr, cp, cy;
+
+    sp = -forward.z;
+
+    float cp_x_cy = forward.x;
+    float cp_x_sy = forward.y;
+    float cp_x_sr = -right.z;
+    float cp_x_cr = up.z;
+
+    float yaw = atan2(cp_x_sy, cp_x_cy);
+    float roll = atan2(cp_x_sr, cp_x_cr);
+
+    cy = cos(yaw);
+    sy = sin(yaw);
+    cr = cos(roll);
+    sr = sin(roll);
+
+    if (fabs(cy) > EPSILON)
+    {
+        cp = cp_x_cy / cy;
+    }
+    else if (fabs(sy) > EPSILON)
+    {
+        cp = cp_x_sy / sy;
+    }
+    else if (fabs(sr) > EPSILON)
+    {
+        cp = cp_x_sr / sr;
+    }
+    else if (fabs(cr) > EPSILON)
+    {
+        cp = cp_x_cr / cr;
+    }
+    else
+    {
+        cp = cos(asin(sp));
+    }
+
+    float pitch = atan2(sp, cp);
+
+    angles[0] = pitch / (M_PI*2.f / 360.f);
+    angles[1] = yaw / (M_PI*2.f / 360.f);
+    angles[2] = roll / (M_PI*2.f / 360.f);
+}
+
+void QuatToYawPitchRoll(XrQuaternionf q, vec3_t rotation, vec3_t out) {
+
+    ovrMatrix4f mat = ovrMatrix4f_CreateFromQuaternion( &q );
+
+    if (rotation[0] != 0.0f || rotation[1] != 0.0f || rotation[2] != 0.0f)
+    {
+        ovrMatrix4f rot = ovrMatrix4f_CreateRotation(radians(rotation[0]), radians(rotation[1]), radians(rotation[2]));
+        mat = ovrMatrix4f_Multiply(&mat, &rot);
+    }
+
+    XrVector4f v1 = {0, 0, -1, 0};
+    XrVector4f v2 = {1, 0, 0, 0};
+    XrVector4f v3 = {0, 1, 0, 0};
+
+    XrVector4f forwardInVRSpace = XrVector4f_MultiplyMatrix4f(&mat, &v1);
+    XrVector4f rightInVRSpace = XrVector4f_MultiplyMatrix4f(&mat, &v2);
+    XrVector4f upInVRSpace = XrVector4f_MultiplyMatrix4f(&mat, &v3);
+
+    XrVector3f forward = {-forwardInVRSpace.z, -forwardInVRSpace.x, forwardInVRSpace.y};
+    XrVector3f right = {-rightInVRSpace.z, -rightInVRSpace.x, rightInVRSpace.y};
+    XrVector3f up = {-upInVRSpace.z, -upInVRSpace.x, upInVRSpace.y};
+
+    XrVector3f forwardNormal = normalizeVec(forward);
+    XrVector3f rightNormal = normalizeVec(right);
+    XrVector3f upNormal = normalizeVec(up);
+
+    GetAnglesFromVectors(forwardNormal, rightNormal, upNormal, out);
+}
+
+//0 = left, 1 = right
+float vibration_channel_duration[2] = {0.0f, 0.0f};
+float vibration_channel_intensity[2] = {0.0f, 0.0f};
```

Comment:
Let's keep all globals at the top of the file to make future refactoring easier (I aim to reduce the amount of globals we use in PPSSPP generally)